README.md
# SimpleSAMLPhp development setup for working with Samly Elixir Library

This is based on `https://github.com/hcpss-banderson/docker-simplesamlphp`
repo. It has been modified to setup SimpleSAMLPhp as an IDP using the
`example-userpass` authentication module. Template for registering a SAML
service provider is also included (`setup/templates/saml20-sp-remote.php.tpl`).

> This is setup to be an Identity Provider (IdP) for a `samly` demo Phoenix
> application ([`samly_howto`](https://github.com/handnot2/samly_howto)).
> But the usage is not limited to `samly_howto`.

## Building Docker Image

```sh
./build.sh
```

This creates `samlysaml:latest` docker image. Modify the `Dockerfile` if you
want to change the `SimpleSAMLPhp` version. The image is based on the stock
Alpine PHP.

## Update `/etc/hosts`

Add the following to your `/etc/hosts` file:

```
# Change the host names appropriately
127.0.0.1 samly.howto idp2.samly.howto idp3.samly.howto
127.0.0.1 idp1.samly idp2.samly idp3.samly
```

## Create a docker-compose wrapper to manage IdP instance

The instructions here help you setup a data container with SimpleSAMLPhp bits, launch PHP
in a separate container as an upstream service and start Nginx fronting
access to PHP.

### Service Provider Application Certificate for SAML communication

Create a self-signed certificate to be used for signed SAML requests. If you are using
Elixir Phoenix based Service provider application, you can use `mix phx.gen.cert`
command to generate a certificate.

Check [`samly_howto`](https://github.com/handnot2/samly_howto) `README.md`
for instructions. This file should be copied to the `setup/sp` directory.

> This certificate is not the same as the key/certificate for your application's
> HTTPS interaction. It should be used only for SAML requests. 

### IdP for use with `samly` configured in `path_segment` model

```sh
#!/bin/sh
# Save this as idp1-compose.sh

export USE_SUBDOMAIN=0
export TIMEZONE=America/New_York # <<- change this if needed

export SP_DOMAIN=samly.howto # <<- change this, this is same as SP HOST in path_segment model
export SP_PORT=4443 # <<- change this
export SP_ENTITY_ID=urn:${SP_DOMAIN}:samly_sp # <<- change this - should match what is in samly config
export SP_CERT_FILE=samly_sp.pem # <<- change this - this should be present in ./setup/sp folder

export PSWD=changeme1 # <<- change this - password for all demo users in ./setup/templates/params.tpl

export IDP_ID=idp1 # <<- change this
export IDP_HOST=${IDP_ID}.samly
export IDP_PORT=9091 # <<- change this

sudo -E ./compose.sh $*
```

Once saved with appropriate updated values, you can use this newly created script to manage the IdP.
This is a simple wrapper around `docker-compose`. You can start IdP using the following command:

```sh
./idp1-compose.sh up -d
./idp1-compose.sh down
./idp1-compose.sh up -d
```

You might be prompted for sudo password.

> You need to `docker-compose down` and `docker-compose up -d` again to
> have a functioning service the first time around. This is due to services
> not knowing when some of the setup is completed before starting. Yes, this
> could be addressed down the line. But this workaround is good enough for now.

You can use the following to get information about the IdP setup with the following command:

```sh
./idp1-compose.sh info
```

This will display something like the following:

```
----------------------------------------------------
IDP Metadata URL = https://idp1.samly:9091/simplesaml/saml2/idp/metadata.php
Assertion Consumer Service URL (ACS) = https://samly.howto:4443/sso/sp/consume/idp1
Single Logout URL (SLO) = https://samly.howto:4443/sso/sp/logout/idp1
SP Base URL = https://samly.howto:4443/sso
SP Entity ID = urn:samly.howto:samly_sp
SP Certificate file = samly_sp.pem
----------------------------------------------------
```

### IdPs for use with `samly` configured in `sub_domain` model

```sh
#!/bin/sh
# Save this as idp2-compose.sh

export USE_SUBDOMAIN=1
export TIMEZONE=America/New_York # <<- change this if needed

export SP_DOMAIN=samly.howto # <<- change this
export SP_PORT=4443 # <<- change this
export SP_ENTITY_ID=urn:${SP_DOMAIN}:samly_sp # <<- change this - should match what is in samly config
export SP_CERT_FILE=samly_sp.pem # <<- change this - this should be present in ./setup/sp folder

export PSWD=changeme2 # <<- change this - password for all demo users in ./setup/templates/params.tpl

export IDP_ID=idp2 # <<- change this
export IDP_HOST=${IDP_ID}.samly
export IDP_PORT=9092 # <<- change this

sudo -E ./compose.sh $*
```

Once saved with appropriate updated values, you can use this newly created script to manage the IdP.
This is a simple wrapper around `docker-compose`. You can start IdP using the following command:

```sh
./idp2-compose.sh up -d
./idp2-compose.sh down
./idp2-compose.sh up -d
```

You might be prompted for sudo password.

You can run the following command to get information about the IdP set with the following command:

```sh
./idp2-compose.sh info
```

This shows out similar to:

```
----------------------------------------------------
IDP Metadata URL = https://idp2.samly:9092/simplesaml/saml2/idp/metadata.php
Assertion Consumer Service URL (ACS) = https://idp2.samly.howto:4443/sso/sp/consume
Single Logout URL (SLO) = https://idp2.samly.howto:4443/sso/sp/logout
SP Base URL = https://idp2.samly.howto:4443/sso
SP Entity ID = urn:samly.howto:samly_sp
SP Certificate file = samly_sp.pem
----------------------------------------------------
```

> You can create another instance of an IdP for use with your sub-domain based Phoenix application
> by simply copying this newly created script as `idp3-compose.sh` and changing the values of
> `IDP_ID`, `IDP_PORT` and password.

## Validate the setup

Visit the URL `https://${IDP_HOST}:${IDP_PORT}/simplesaml`. You should be able to login
as administrator using the password you set in your compose script.

Check out the `Federation` tab. You should see the SAML service provider.

If there are any errors in the process, you should be able to change the
variables in your compose script file, bring down the containers
using `./idp1-compose.sh down` and re-create using `./idp1-compose.sh up -d`.

## Caution

> Keep in mind this is meant for development purposes only.


CHANGELOG.md
# Changelog

v0.5

* `README.md` updates

v0.4

* Switch to SimpleSAMLphp version 1.16.3
* Enable HTTPS
* Use specific version of Docker php base image (issue #3)
* Enable assertion encryption
* Support multi-valued roles for users in authsource
* Support multiple IdPs (check README.md)

v0.3

* Switch to SimpleSAMLphp version 1.16.2

v0.2

* Fixed TLS error for wget
  PR from [brianmay](https://github.com/brianmay)

v0.1

* Added README.md and CHANGELOG.md files


setup/init.sh
#!/bin/sh

SRC_DIR="/setup"
GEN_DIR="/sspdv"

SSP_DIR="/srv/simplesaml"

mkdir -p ${GEN_DIR}/cert
mkdir -p ${GEN_DIR}/nginx/conf.d
touch ${GEN_DIR}/nginx/conf.d/site.conf

if [ ! -f ${GEN_DIR}/cert/server.crt ];
then
  C="US"
  ST="Midlands"
  L="Safeville"
  O="ID Federation"
  OU="Department of Identities"
  CN="${IDP_HOST}"
  SUBJ="/C=${C}/ST=${ST}/L=${L}/O=${O}/OU=${OU}/CN=${CN}"

  echo "Generating IDP Certificate ..."
  openssl req -new -x509 -sha256 -days 365 -nodes \
    -newkey rsa:4096 \
    -out ${GEN_DIR}/cert/server.crt \
    -keyout ${GEN_DIR}/cert/server.pem \
    -subj "${SUBJ}"
fi

if [ -d ${SRC_DIR}/sp ];
then
  echo "Copying SP cert ..."
  cp -r ${SRC_DIR}/sp ${SSP_DIR}/cert/
fi
cp -f ${GEN_DIR}/cert/server.crt ${SSP_DIR}/cert/
cp -f ${GEN_DIR}/cert/server.pem ${SSP_DIR}/cert/

echo > ${GEN_DIR}/${IDP_ID}_starter.yml
echo "idp_id: ${IDP_ID}"                                >> ${GEN_DIR}/${IDP_ID}_starter.yml
echo "idp_host: ${IDP_HOST}"                            >> ${GEN_DIR}/${IDP_ID}_starter.yml
echo "idp_port: ${IDP_PORT}"                            >> ${GEN_DIR}/${IDP_ID}_starter.yml
echo "secret_salt: ${SECRET_SALT}"                      >> ${GEN_DIR}/${IDP_ID}_starter.yml
echo "pswd: ${PSWD}"                                    >> ${GEN_DIR}/${IDP_ID}_starter.yml
echo "sp_baseurl: ${SP_BASEURL}"                        >> ${GEN_DIR}/${IDP_ID}_starter.yml
echo "sp_acs_uri: ${SP_ACS_URI}"                        >> ${GEN_DIR}/${IDP_ID}_starter.yml
echo "sp_slo_uri: ${SP_SLO_URI}"                        >> ${GEN_DIR}/${IDP_ID}_starter.yml
echo "sp_entity_id: ${SP_ENTITY_ID}"                    >> ${GEN_DIR}/${IDP_ID}_starter.yml
echo "sp_cert_file: ${SSP_DIR}/cert/sp/${SP_CERT_FILE}" >> ${GEN_DIR}/${IDP_ID}_starter.yml
echo "timezone: ${TIMEZONE}"                            >> ${GEN_DIR}/${IDP_ID}_starter.yml

twit ${SRC_DIR}/templates/params.tpl ${GEN_DIR}/${IDP_ID}_params.yml \
  -p ${GEN_DIR}/${IDP_ID}_starter.yml -n

for i in `cat ${SRC_DIR}/templates/nginx.config`
do
  f=`echo $i | sed 's/\.tpl$//' -`
  twit ${SRC_DIR}/templates/${i} ${GEN_DIR}/nginx/conf.d/${f} \
    -p ${SRC_DIR}/params/defaults.yml -p ${GEN_DIR}/${IDP_ID}_params.yml -n
done

for i in `cat ${SRC_DIR}/templates/ssp.config`
do
  f=`echo $i | sed 's/\.tpl$//' -`
  twit ${SRC_DIR}/templates/${i} ${SSP_DIR}/config/${f} \
    -p ${SRC_DIR}/params/defaults.yml -p ${GEN_DIR}/${IDP_ID}_params.yml -n
done

for i in `cat ${SRC_DIR}/templates/ssp.metadata`
do
  f=`echo $i | sed 's/\.tpl$//' -`
  twit ${SRC_DIR}/templates/${i} ${SSP_DIR}/metadata/${f} \
    -p ${SRC_DIR}/params/defaults.yml -p ${GEN_DIR}/${IDP_ID}_params.yml -n
done

touch ${SSP_DIR}/modules/exampleauth/default-enable

rm -rf ${GEN_DIR}/${IDP_ID}_starter.yml
rm -rf ${GEN_DIR}/${IDP_ID}_params.yml


compose.sh
#!/bin/sh
#
# =============================================================================
# USE_SUBDOMAIN: 0 - if your application is not multi-tenant enabled
#                1 - if multi-tenant enabled (t1.mydomain, t2.mydomain etc)
# SP_DOMAIN: set to your application domain (mydomain) when USE_SUBDOMAIN=1
#            set to hostname where your app is deployed when USE_SUBDOMAIN=0
# SP_PORT: Port where your application is listening
# SP_ENTITY_ID: Typically of the form "urn:myapp:sp1". This should be same as
#               the value specified in samly service_providers config
# SP_CERT_FILE: Just the file name - NOT the path. This file MUST be placed in
#               ./setup/sp folder prior to running this script
# PSWD: Password for the demo users (check setup/templates/params.tpl)
#
# IDP_ID: This is same as the id used in samly identity_providers config
# IDP_HOST: Host name for IdP
# IDP_PORT: Where IdP listens.
#
# TIMEZONE: SimpleSAMLphp uses this in logs.
# =============================================================================

export USE_SUBDOMAIN="${USE_SUBDOMAIN:-0}"

SP_HOST=samly.howto
export SP_DOMAIN="${SP_DOMAIN:-${SP_HOST}}"
export SP_PORT="${SP_PORT:-4443}"

DEFAULT_ENTITY_ID="urn:${SP_DOMAIN}:sp"
export SP_ENTITY_ID="${SP_ENTITY_ID:-${DEFAULT_ENTITY_ID}}"

export SP_CERT_FILE="${SP_CERT_FILE:-sp.crt}"

export PSWD="${PSWD:-changeme}"

export TIMEZONE="${TIMEZONE:-America/New_York}"

export IDP_ID="${IDP_ID:-idp1}"
export IDP_HOST="${IDP_HOST:-${IDP_ID}.samly}"
export IDP_PORT="${IDP_PORT:-9091}"

export SECRET_SALT=`tr -c -d '0123456789abcdefghijklmnopqrstuvwxyz' </dev/urandom | dd bs=32 count=1 2>/dev/null`

if [ "${USE_SUBDOMAIN}" = "1" ]; then
  export SP_BASEURL="https://${IDP_ID}.${SP_DOMAIN}:${SP_PORT}/sso"
  export SP_ACS_URI="/sp/consume"
  export SP_SLO_URI="/sp/logout"
else
  export SP_BASEURL="https://${SP_DOMAIN}:${SP_PORT}/sso"
  export SP_ACS_URI="/sp/consume/${IDP_ID}"
  export SP_SLO_URI="/sp/logout/${IDP_ID}"
fi

if [ "$1" = "info" ]; then
  echo "----------------------------------------------------"
  echo "IDP Metadata URL = https://${IDP_HOST}:${IDP_PORT}/simplesaml/saml2/idp/metadata.php"
  echo "Assertion Consumer Service URL (ACS) = ${SP_BASEURL}${SP_ACS_URI}"
  echo "Single Logout URL (SLO) = ${SP_BASEURL}${SP_SLO_URI}"
  echo "SP Base URL = ${SP_BASEURL}"
  echo "SP Entity ID = ${SP_ENTITY_ID}"
  echo "SP Certificate file = ${SP_CERT_FILE}"
  echo "----------------------------------------------------"

  exit 0
else
  docker-compose -p "${IDP_ID}" $*
fi
